### PR TITLE
Implement collective gather op

### PR DESCRIPTION
**Pull request by bfolie**
#9315

### CODE REVIEW

File: test/pjrt/test_collective_ops_tpu.py [hunk 1]
```diff
@@ -103,29 +103,6 @@ def test_reduce_scatter(self, pin_layout):
...
-    dist.scatter(output_tensor, tensors, src=0)
-    return output_tensor.cpu()
-
-  def test_scatter(self):
-    """self._scatter instantiates a list of tensors [[0], [1], ..., [n-1]]
-    on device 0, then scatters it. Device i should therefore receive [i]."""
-    results = pjrt.run_multiprocess(self._scatter)
-    for ordinal, value in results.items():
-      np.testing.assert_array_equal(value, [ordinal])
-
```
Comment (bfolie) [inline test/pjrt/test_collective_ops_tpu.py]: Just moving this test into the appropriate class

File: test/pjrt/test_collective_ops_tpu.py [hunk 2]
```diff
@@ -359,6 +336,82 @@ def test_all_to_all_single(self, use_dynamo):
...
+    output_tensor = torch.tensor([-1], dtype=torch.float, device=device)
+    dist.scatter(output_tensor, tensors, src=0)
+    return output_tensor.cpu()
+
+  def test_scatter(self):
+    """self._scatter instantiates a list of tensors [[0], [1], ..., [n-1]]
+    on device 0, then scatters it. Device i should therefore receive [i]."""
+    results = pjrt.run_multiprocess(self._scatter)
+    for ordinal, value in results.items():
+      np.testing.assert_array_equal(value, [ordinal])
```
Comment (bfolie) [inline test/pjrt/test_collective_ops_tpu.py]: copied from above

**Comment by bfolie** [comment]
Failing tests are expected until the TPU CI cluster is updated to use python 3.12. See #9434

File: torch_xla/distributed/xla_backend.py [hunk 3]
```diff
@@ -250,8 +250,44 @@ def alltoall_base(self, output, input, output_split_sizes, input_split_sizes,
...
+    rank = xr.global_ordinal()
+
+    for i, input_tensor in enumerate(input_tensor_list):
+      is_scalar = input_tensor.dim() == 0
+      input_for_all_gather = (
+          input_tensor.clone().reshape(1) if is_scalar else input_tensor)
+
+      gathered_tensor = xm.all_gather(
+          input_for_all_gather, dim=0, groups=self._mesh, pin_layout=False)
+      # Syncing is required to keep the heterogeneous copying below at the
```
Comment (pgmoka) [inline torch_xla/distributed/xla_backend.py]: NIT: Add space between code line and comment.

File: torch_xla/distributed/xla_backend.py [hunk 4]
```diff
@@ -250,8 +250,44 @@ def alltoall_base(self, output, input, output_split_sizes, input_split_sizes,
...
+  # Input tensors are gathered into list of output tensors on the dst device.
+  # Output tensors list is None for all non-dst devices.
+  # This is an inefficient operation. In order to avoid XLA deadlocks it
+  # performs redundant gathers on non-dst devices and materializes the result.
+  def gather(self, output_tensors_list: list[list[torch.Tensor]],
+             input_tensor_list: list[torch.Tensor], opts: GatherOptions):
+    rank = xr.global_ordinal()
+
+    for i, input_tensor in enumerate(input_tensor_list):
+      is_scalar = input_tensor.dim() == 0
```
Comment (pgmoka) [inline torch_xla/distributed/xla_backend.py:294]: This is happening during each iteration of the loop.

If input_tensor_list is not empty, could we not do something like `is_scalar = input_tensor_list[0].dim() == 0`?

File: torch_xla/distributed/xla_backend.py [hunk 5]
```diff
@@ -250,8 +250,44 @@ def alltoall_base(self, output, input, output_split_sizes, input_split_sizes,
...
+          gathered_chunks = torch.split(gathered_tensor, chunk_size, dim=0)
+          for j, chunk in enumerate(gathered_chunks):
+            if chunk.shape != output_tensors[j].shape:
+              chunk = chunk.reshape(output_tensors[j].shape)
+            output_tensors[j].copy_(chunk)
+
+    if rank == opts.rootRank:
+      return _ret_work(output_tensors_list)
+    else:
+      return _ret_work([[]])
```
Comment (pgmoka) [inline torch_xla/distributed/xla_backend.py:321]: What is going on here? From this base reading, it is that if `rank != opts.rootRank`, return an empty.

If that is the case, could we add something like:

```
if rank != opts.rootRank:
      return _ret_work([[]])
```

In the beginning of the function, and avoid this if split, as well as the one above?

Comment (bfolie) [reply torch_xla/distributed/xla_backend.py:321]: Good point -- that would make the code simpler

File: torch_xla/distributed/xla_backend.py (hunk 4 above)
Comment (bfolie) [reply torch_xla/distributed/xla_backend.py:294]: It's not guaranteed that every element of `input_tensor_list` has the same size. They're basically independent gather operations.

File: torch_xla/distributed/xla_backend.py (hunk 5 above)
Comment (bfolie) [reply torch_xla/distributed/xla_backend.py:321]: Actually no -- all non-dst ranks still need to call the all_gather with their input and then sync. It's only the copying to the output which is device-specific. Which means we can't return at the beginning of the function

File: torch_xla/distributed/xla_backend.py (hunk 4 above)
Comment (pgmoka) [reply torch_xla/distributed/xla_backend.py:294]: Ah! Gotcha

File: torch_xla/distributed/xla_backend.py (hunk 5 above)
Comment (pgmoka) [reply torch_xla/distributed/xla_backend.py:321]: I see. On that case, I think this is a reasonable enough state.

**Review by pgmoka — APPROVED** [review]
